File: dertet-harness-ios/DertetHarness/Network/OpenAICompatibleClient.swift
import Foundation

final class OpenAICompatibleClient: ChatClient {
    func streamChat(
        baseUrl: String,
        apiKey: String,
        model: String,
        systemPrompt: String,
        history: [ChatMessage]
    ) -> AsyncThrowingStream<ChatStreamEvent, Error> {
        AsyncThrowingStream { continuation in
            let task = Task {
                var gotAny = false
                do {
                    var messages: [[String: Any]] = []
                    if !systemPrompt.isEmpty {
                        messages.append(["role": "system", "content": systemPrompt])
                    }
                    for msg in history {
                        messages.append(["role": msg.role, "content": buildContent(msg)])
                    }

                    let body: [String: Any] = ["model": model, "stream": true, "messages": messages]

                    guard let url = URL(string: baseUrl.trimmedTrailingSlash + "/chat/completions") else {
                        continuation.finish(throwing: URLError(.badURL))
                        return
                    }
                    var request = URLRequest(url: url)
                    request.httpMethod = "POST"
                    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
                    request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
                    if baseUrl.contains("openrouter.ai") {
                        request.setValue("https://dertetharness.app", forHTTPHeaderField: "HTTP-Referer")
                        request.setValue("Dertet Harness", forHTTPHeaderField: "X-Title")
                    }
                    request.httpBody = jsonString(body)

                    for try await event in sseEvents(for: request) {
                        try Task.checkCancellation()
                        let data = event.data.trimmingCharacters(in: .whitespaces)
                        if data.isEmpty || data == "[DONE]" { continue }
                        guard let jsonData = data.data(using: .utf8),
                              let obj = try? JSONSerialization.jsonObject(with: jsonData) as? [String: Any] else { continue }

                        if let errorObj = asDict(obj["error"]), let msg = asString(errorObj["message"]) {
                            continuation.yield(.error(msg))
                            continue
                        }
                        if let choices = asArray(obj["choices"]),
                           let first = asDict(choices.first),
                           let delta = asDict(first["delta"]),
                           let text = asString(delta["content"]), !text.isEmpty {
                            gotAny = true
                            continuation.yield(.delta(text))
                        }
                    }
                    continuation.yield(.done)
                    continuation.finish()
                } catch {
                    if !gotAny {
                        continuation.yield(.error(error.localizedDescription))
                    }
                    continuation.finish()
                }
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }

    private func buildContent(_ msg: ChatMessage) -> Any {
        let images = msg.attachments.filter { $0.kind == .image && $0.base64Data != nil }
        let files = msg.attachments.filter { $0.kind == .file }

        var text = msg.text
        for f in files where f.textContent != nil {
            text = "[File: \(f.fileName)]\n\(f.textContent!)\n\n\(text)"
        }

        if images.isEmpty { return text }

        var parts: [[String: Any]] = [["type": "text", "text": text]]
        for img in images {
            parts.append([
                "type": "image_url",
                "image_url": ["url": "data:\(img.mimeType);base64,\(img.base64Data!)"]
            ])
        }
        return parts
    }
}

extension String {
    var trimmedTrailingSlash: String {
        hasSuffix("/") ? String(dropLast()) : self
    }
}

File: dertet-harness-android/app/src/main/java/com/dertet/harness/network/OpenAiCompatibleClient.kt
package com.dertet.harness.network

import com.dertet.harness.data.AttachmentKind
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

class OpenAiCompatibleClient : ChatClient {

    override fun streamChat(
        baseUrl: String,
        apiKey: String,
        model: String,
        systemPrompt: String,
        history: List<ChatMessage>
    ): Flow<ChatStreamEvent> = flow {
        val messages = buildJsonArray {
            if (systemPrompt.isNotBlank()) {
                add(buildJsonObject {
                    put("role", "system")
                    put("content", systemPrompt)
                })
            }
            for (msg in history) {
                add(buildJsonObject {
                    put("role", msg.role)
                    put("content", buildContent(msg))
                })
            }
        }

        val body = buildJsonObject {
            put("model", model)
            put("stream", true)
            put("messages", messages)
        }.toString().toRequestBody("application/json".toMediaType())

        val url = baseUrl.trimEnd('/') + "/chat/completions"
        val requestBuilder = Request.Builder()
            .url(url)
            .addHeader("Authorization", "Bearer $apiKey")
            .addHeader("Content-Type", "application/json")
        if (baseUrl.contains("openrouter.ai")) {
            requestBuilder.addHeader("HTTP-Referer", "https://dertetgpt.app")
            requestBuilder.addHeader("X-Title", "Dertet Harness")
        }
        val request = requestBuilder.post(body).build()

        var gotAny = false
        try {
            sseFlow(sharedHttpClient, request).collect { event ->
                val data = event.data.trim()
                if (data == "[DONE]") return@collect
                if (data.isEmpty()) return@collect
                try {
                    val json = Json.parseToJsonElement(data).jsonObject
                    val errorObj = json["error"]?.jsonObject
                    if (errorObj != null) {
                        val msg = errorObj["message"]?.jsonPrimitive?.contentOrNull ?: "API error"
                        emit(ChatStreamEvent.Error(msg))
                        return@collect
                    }
                    val choices = json["choices"]?.jsonArray ?: JsonArray(emptyList())
                    val delta = choices.firstOrNull()?.jsonObject?.get("delta")?.jsonObject
                    val content = delta?.get("content")
                    val text = if (content == null || content is JsonNull) null else content.jsonPrimitive.contentOrNull
                    if (!text.isNullOrEmpty()) {
                        gotAny = true
                        emit(ChatStreamEvent.Delta(text))
                    }
                    val images = delta?.get("images")?.jsonArray
                    val imageUrl = images?.firstOrNull()?.jsonObject
                        ?.get("image_url")?.jsonObject
                        ?.get("url")?.jsonPrimitive?.contentOrNull
                    if (!imageUrl.isNullOrEmpty()) {
                        parseDataUrl(imageUrl)?.let { (mime, data) ->
                            gotAny = true
                            emit(ChatStreamEvent.ImageGenerated(mime, data))
                        }
                    }
                } catch (e: Exception) {
                    // ignore malformed keep-alive lines
                }
            }
            emit(ChatStreamEvent.Done)
        } catch (e: Exception) {
            if (!gotAny) emit(ChatStreamEvent.Error(e.message ?: "Помилка з’єднання"))
            else emit(ChatStreamEvent.Done)
        }
    }

    private fun buildContent(msg: ChatMessage): kotlinx.serialization.json.JsonElement {
        val images = msg.attachments.filter { it.kind == AttachmentKind.IMAGE && it.base64Data != null }
        val files = msg.attachments.filter { it.kind == AttachmentKind.FILE }

        var textPart = msg.text
        for (f in files) {
            if (f.textContent != null) {
                textPart = "[Файл: ${f.fileName}]\n${f.textContent}\n\n$textPart"
            }
        }

        if (images.isEmpty()) {
            return kotlinx.serialization.json.JsonPrimitive(textPart)
        }

        return buildJsonArray {
            add(buildJsonObject {
                put("type", "text")
                put("text", textPart)
            })
            for (img in images) {
                add(buildJsonObject {
                    put("type", "image_url")
                    put("image_url", buildJsonObject {
                        put("url", "data:${img.mimeType};base64,${img.base64Data}")
                    })
                })
            }
        }
    }
}
